src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -422,20 +420,24 @@ - (void)writeAttributeWithEndpointID:(NSNumber *)endpointID
                expectedValueInterval:(NSNumber *)expectedValueInterval
                    timedWriteTimeout:(NSNumber * _Nullable)timeout
 {
-    // Start the asynchronous operation
-    MTRBaseDevice * baseDevice = [self newBaseDevice];
-    [baseDevice
-        writeAttributeWithEndpointID:endpointID
-                           clusterID:clusterID
-                         attributeID:attributeID
-                               value:value
-                   timedWriteTimeout:timeout
-                               queue:self.queue
-                          completion:^(NSArray<NSDictionary<NSString *, id> *> * _Nullable values, NSError * _Nullable error) {
-                              if (values) {
-                                  [self _handleAttributeReport:values];
-                              }
-                          }];
+    MTRAsyncCallbackQueueWorkItem * workItem = [[MTRAsyncCallbackQueueWorkItem alloc] initWithQueue:_queue];
+    MTRAsyncCallbackReadyHandler readyHandler = ^(MTRDevice * device, NSUInteger retryCount) {
+        MTRBaseDevice * baseDevice = [self newBaseDevice];
+        [baseDevice
+            writeAttributeWithEndpointID:endpointID
+                               clusterID:clusterID
+                             attributeID:attributeID
+                                   value:value
+                       timedWriteTimeout:timeout
+                                   queue:self.queue
+                              completion:^(NSArray<NSDictionary<NSString *, id> *> * _Nullable values, NSError * _Nullable error) {
+                                  if (values) {
+                                      [self _handleAttributeReport:values];
```

Comment:
Pre-existing, but this seems strange now that I look at it.  The only reason `values` would be nonempty here is if there was a path-specific error on the write, which is being communicated in `values`.  But then treating that as an attribute _report_ is a bit odd...  Need to figure out how the error reporting here should actually work.

Reply:
Filed #23285 to investigate